### ISSUE TITLE
Migration hooks for modifying data

### ISSUE THREAD

**Issue by sarink** [prisma/prisma#15384]
## Problem

Prisma migrate only provides a way to modify the db schema, which is insufficient for building an application. Expecting users to write raw sql to modify data and/or handle data migrations manually is undesirable (and a bit of a surprise ask from an orm + migration tool).

## Suggested solution

It seems like it would be quite easy for prisma to at least add pre/post migration hooks, allowing the user to supply, eg a `prisma/migrations/timestamp_name/before.js` file which will run before `migration.sql`, and a `prisma/migrations/timestamp_name/after.js` file which will run after `migration.sql`.

## Alternatives

Modifying large amounts of data purely in a SQL by manually modifying `migration.sql` is unacceptable

**Comment by DavidHancu** [comment]
👋 **Hi @sarink!**

I've been working on my NPM Package [Prisma Util](https://www.npmjs.com/package/prisma-util), which is all about implementing features that are requested by developers in the community. This issue has attracted my attention and I want to ask if you'd be interested in it being implemented in Prisma Util.

Would you use this feature if it was added to Prisma Util?

**Comment by JVMartin** [comment]
@DavidHancu I think I speak for everyone when I say, no, an ORM should run migrations via an ORM. All modern ORMs do this. To be limited to raw SQL is simply a design mistake.

**Comment by JVMartin** [comment]
This is a duplicate of #11810 .

I don't believe hooks are the answer. The entire migration itself should be able to use the ORM, you can execute raw queries via the ORM if you want to. No need for hooks.

**Comment by giovanioliani** [comment]
I'm still confused about whether or not pure SQL data migrations are even supported. I modified a migration that changes the schema to have data processing in between DDL changes (set not null, etc) and the migration fails/hangs half the time. I don't even need JS/prisma client support, I just want it to properly run the SQL file I gave it. 

Does anyone know if this is supported? I only got silence for an answer on their discord server.

```
-- DropForeignKey
ALTER TABLE "BackOrderQueue" DROP CONSTRAINT "BackOrderQueue_inventoryItemId_fkey";

-- DropIndex
DROP INDEX "BackOrderQueue_inventoryItemId_orgId_key";

-- add sku column
ALTER TABLE "BackOrderQueue" ADD COLUMN "sku" TEXT;

-- set skus from corresponding items on backorder table(1->1)
update "BackOrderQueue" bo
set sku = ii.sku
from "InventoryItem" ii 
inner join "Store" s on s.id = ii."storeId" 
where ii.id = bo."inventoryItemId" and s."authProviderId" = bo."orgId";

-- set non null constraint on BO sku since every row should have a sku now
ALTER TABLE "BackOrderQueue" ALTER COLUMN sku SET NOT NULL;

-- sets total quantity of each item on backorder queue. This will leave "duplicate" columns with same sku, quantity but different inventoryItemId
update "BackOrderQueue" bo
set quantity = innquery.qty
from (
	-- select total stock for each sku, store pair
	select sum(quantity) qty , sku, s.id as "storeId"
	from "BackOrderQueue" boq
	inner join "Store" s on s."authProviderId" = boq."orgId"
	group by sku, s.id
) innquery
inner join "Store" s on s.id = innquery."storeId"
inner join "InventoryItem" ii on ii."storeId"  = s.id
where bo."orgId" = s."authProviderId" and ii.id = bo."inventoryItemId"
	  and bo.sku = innquery.sku;

--delete repeated rows
delete from "BackOrderQueue" where id in (
	-- select ids from repeated rows
	select id from (
		-- get ids and assign numbers to org,sku pairs, with repeated pairs having numbers > 1
		select id, row_number() over(partition by "orgId", sku) as row_num from "BackOrderQueue"
	) innerquery where row_num > 1
);

-- AlterTable
ALTER TABLE "BackOrderQueue" DROP COLUMN "inventoryItemId";

-- CreateIndex
CREATE UNIQUE INDEX "BackOrderQueue_sku_orgId_key" ON "BackOrderQueue"("sku", "orgId");

```

**Comment by Softjey** [comment]
Hi everyone! 👋

I’ve developed a small library to help address this problem: [Prisma Data Migrations](https://www.npmjs.com/package/prisma-data-migrations?activeTab=readme). This library enhances Prisma’s migration system by enabling the use of post-migration scripts for data migrations alongside schema changes. And allows migrate to specific migration using CLI.

You can find the repository here: [GitHub Repository](https://github.com/Softjey/prisma-data-migrations).

Feel free to open issues, report bugs, or suggest improvements. I’m also open to pull requests if you’d like to contribute to the project!

Looking forward to your feedback! 🚀

**Comment by JVMartin** [comment]
That was not necessary @Softjey , [Prisma has first class support for data migrations now](https://www.prisma.io/docs/guides/data-migration-with-expand-and-contract), and it's perfectly implemented.

**Comment by Softjey** [comment]
> That was not necessary @Softjey , [Prisma has first class support for data migrations now](https://www.prisma.io/docs/guides/data-migration-with-expand-and-contract), and it's perfectly implemented.

@JVMartin 

I encountered this situation in my project as well, and initially, I tried using the documented approach. However, I ran into several challenges when managing multiple migrations and data migrations in a single release. Here's an overview of the issues:

1. Multiple migrations in a single release: For example, I needed to apply the following sequence in one release:

Add countries table

Populate countries table (data migration)

Add city model

Populate cities table (data migration)

Make some schema changes

Apply additional data migrations


All these changes had to be applied together as part of a single deployment. Applying them step-by-step wasn't feasible due to the deployment strategy.


2. Multiple environments: Our workflow included four environments (development, staging, testing, production). Applying the migrations manually across all environments became tedious and error-prone, especially when handling data migrations in a strict sequence.


3. Data migrations and seed duplication: Every data migration script had to be duplicated and kept up-to-date in seed files, as the schema kept evolving. This duplication not only increased the maintenance overhead but also made it more challenging to ensure consistency between migrations and seeds.



To address these challenges, we developed a library that simplifies the management of schema and data migrations in complex environments. It streamlines the process of running multiple migrations, ensures consistency across environments, and eliminates the need for seed duplication by integrating schema and data management.

The library worked so well for us that we decided to refine it further and release it as an open-source project.

**Comment by JVMartin** [comment]
Nobody is manually applying migrations. We run them automatically on deployment in every environment. The Prisma built-in data migrations run automatically after the corresponding schema migration. We can bundle as many schema and data migrations in a single release as we want.

**Comment by Stadly** [comment]
> Nobody is manually applying migrations. We run them automatically on deployment in every environment. The Prisma built-in data migrations run automatically after the corresponding (optional) schema migration. We can bundle as many schema and data migrations in a single release as we want.
> 
> They show you how to do it manually the guide but it's completely trivial to automate it, which is what everyone does.

Automating it doesn't seem trivial to me. Can you elaborate?

**Comment by JVMartin** [comment]
@Stadly Certainly you aren't running `npx prisma migrate` by hand in any environments, right? I've worked with 5 different organizations that use Prisma extensively, and all of us simply bundle migrations into the commit, and the migrate script is run automatically on application bootstrap in every environment. I don't know a single company that ever runs migrations by hand. The same is true regardless of whether they are schema or data migrations.

**Comment by JVMartin** [comment]
Here is an example PrismaService from one of our platforms:

```typescript
@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit {
  constructor(
    @InjectPinoLogger(PrismaService.name)
    private readonly logger: Logger,
    private readonly configService: ConfigService,
  ) {
    super({
      ...(configService.logQueries
        ? {
            log: [
              {
                emit: 'event',
                level: 'query',
              },
            ],
          }
        : {}),
    });
  }

  async onModuleInit(): Promise<any> {
    // In production-like environments (docker + k8s), only the scheduler runs migrations / does database sync.
    // In development environment (API only), the API runs migrations / does database sync.
    if (
      (this.configService.nodeEnv === 'production' && this.configService.type === 'scheduler') ||
      (this.configService.nodeEnv === 'development' && this.configService.type === 'api')
    ) {
      let command = ['production', 'qa'].includes(this.configService.env)
        ? 'npx prisma migrate deploy'
        : 'npx prisma db push --accept-data-loss';

      if (this.configService.nodeEnv === 'development') {
        command += ' --schema prisma-sqlite/schema.prisma';
      }

      const { stdout, stderr } = await exec(command, {
        cwd: PROJECT_PATH,
        env: process.env,
      });
      this.logger.info(stdout);
      if (stderr) {
        this.logger.error(stderr);
      }
    }

    await this.$connect();

    (this.$on as any)('query', (e) => {
      this.logger.info({ query: e.query, params: e.params, duration: e.duration }, `query`);
    });

    // TODO:
    // Fix if they ever make extensions not disgustingly bad https://github.com/prisma/prisma/issues/18628
    this.$use(async (params, next) => {
      if (params.model === 'User' && params.action === 'create') {
        const user = await next(params);
        const hashid = hashidsService.encode(user.id);
        return await this.user.update({
          where: { id: user.id },
          data: { hashid },
        });
      }

      return next(params);
    });
  }
}
```

**Comment by Stadly** [comment]
> @Stadly Certainly you aren't running `npx prisma migrate` by hand in any environments, right? I've worked with 5 different organizations that use Prisma extensively, and all of us simply bundle migrations into the commit, and the migrate script is run automatically on application bootstrap in every environment. I don't know a single company that ever runs migrations by hand. The same is true regardless of whether they are schema or data migrations.

The problem is that `npx prisma migrate` does not handle data migration (unless part of the `.sql` migration file, which is not always feasible). How can you automate running the data migration command (`tsx ./prisma/migrations/<migration-timestamp>/data-migration.ts` in the link you posted) along with the schema migration?

As far as I can tell, the code example you posted does not automate data migration, only schema migration.

**Comment by JVMartin** [comment]
Toss in a migration tracking table:
```prisma
model DataMigrationHistory {
  id             Int      @id @default(autoincrement()) // Primary key
  migrationName  String   @unique                     // Name of the migration
  executedAt     DateTime @default(now())             // Timestamp when migration was executed
}
```

Add to your `onModuleInit()`:
```typescript
async function runMigrations() {
  for (const migration of migrations) {
    const existingMigration = await prisma.dataMigrationHistory.findFirst({
      where: { migrationName: migration.name },
    });

    if (!existingMigration) {
      console.log(`Running migration: ${migration.name}`);
      try {
        await migration.execute();
        await prisma.dataMigrationHistory.create({
          data: { migrationName: migration.name },
        });
        console.log(`Completed migration: ${migration.name}`);
      } catch (err) {
        console.error(`Failed migration: ${migration.name}`, err);
        process.exit(1); // Exit on error
      }
    } else {
      console.log(`Skipping migration (already run): ${migration.name}`);
    }
  }
}
```

So you always run all data migrations from your data migrations folder that haven't been ran yet. You can decide whether to do it before, after, or between data migrations, etc.

**Comment by JVMartin** [comment]
@Softjey In your package, you want them to copy the entire schema.prisma into each migration? Isn't that a ton of duplicated code?

**Comment by DavidHancu** [comment]
Probably because they generate a separate client

**Comment by JVMartin** [comment]
So if your schema.prisma is 800 lines long, and you have 25 data migrations, that's now 20,000 lines of Prisma schema in your repo? 😕

**Comment by Softjey** [comment]
> So if your schema.prisma is 800 lines long, and you have 25 data migrations, that's now 20,000 lines of Prisma schema in your repo? 😕

I agree that this is far from ideal, and I’ve been thinking about how to address it. It would be great to have the ability to restore the `schema.prisma` file based on the existing migrations. However, as far as I understand, Prisma doesn’t currently offer such functionality. At this stage of the library’s development, I’ve decided not to implement this feature myself. 

In such a case, we could store Prisma schemas in `node_modules` and simply generate them using the `generate` command, much like how types are handled. If you have a solution to this problem, I’d be happy to discuss it further.

Additionally, I’ll soon be adding a command specifically for users who work with `multiSchema`. This will enable merging everything into a single schema, streamlining the process.

**Comment by JVMartin** [comment]
It sounds like you're trying to solve the problem where people want to bundle into a single commit:
1. Data migration
2. Schema migration
3. Data migration
4. Schema migration

But I think there's a reasonable argument to be said that this is an anti-pattern. Not the least of which, because even if you've tested on a fork of production, data migrations are soo easy to fail when data changes out from underneath you between the data integration test and the actual deployment.

If you simply require that data migrations always use the latest Prisma schema, you can eliminate the need to include multiple Prisma schemas in the repo.

In other words, all outstanding schema migrations must run to completion successfully first. Then, apply all outstanding data migrations.

**Comment by Softjey** [comment]
I agree that this might look like an anti-pattern, but the use case I had required approximately four `schema -> data` migrations, all of which logically needed to be part of a single release. I considered restructuring the sequence to apply schema changes first and then the data, but in my case, this approach was not feasible. 

Instead, I decided to test everything thoroughly on database replicas and then apply the mechanism described in the library. 

To ensure atomicity in the future, I plan to combine the schema migration and post-migration script into a single transaction. This would help mitigate the issue of instability in data migrations.

**Comment by JVMartin** [comment]
It's certainly very cool that you were able to pull that off, and it's an interesting approach to the problem.

I also need to revise my earlier statement, I think we could eliminate duplicate Prisma schemas, and still support:

1. Data migration
2. Schema migration
3. Data migration
4. Schema migration

The real simplifying constraint would be that all the data migrations would have to work with the latest / current Prisma schema.

In other words, it would be fine to interleave data migrations with schema migrations, but since the data migrations are all currently using the current Prisma schema (in this de-duplicated proposal), if you're for example referencing a column that doesn't exist anymore in the latest schema, you might have to e.g. use raw queries in just those rare cases.

I've done some heinously complex migrations over the years for example with TypeORM that has native data migrations that work this way, and I've never found a situation where I couldn't accomplish it, and it was exceptionally rare that I ever needed to use a raw query.

**Comment by Softjey** [comment]
Yes, this would reduce the number of duplicated schemas, but it could potentially cause issues when switching to a different client in the future, making the post-migration script uncompilable. For example, consider a data migration where we use something like `prisma.user.updateMany(...)`. If, at some point in the future, we rename the `User` table to `Account`, the expression `prisma.user.updateMany(...)` would no longer compile in TypeScript. 

I’ve considered writing a script that could regenerate the `schema.prisma` file based on schema migrations, which would eliminate the need to manually carry it over. However, I haven’t implemented this yet. 

If you have any suggestions or ideas on how such a script could be easily written, I’m open to hearing your proposals!

**Comment by JVMartin** [comment]
I hear what you're saying -- but I'd say for 99.5% of us, data migrations are *only* for production. All other environments are seeded using current Prisma schema. So there's never a situation where you'd ever run a data migration except the moment you merge it to master / main.

Do you really fall int hat 0.5% where you have multiple production environments running random data migrations from distant past points in time? 😅

**Comment by Softjey** [comment]
Unfortunately, this was exactly the case for me, as we have two production environments: one for beta testing and one for the main environment. Over the course of two commits that were deployed to beta but not yet to the main environment, the schema evolved in a way that caused type conflicts.

**Comment by JVMartin** [comment]
I gotcha. Now that I understand your very niche / nuanced use case, your including of Prisma schemas with each data migration makes perfect sense. 👍 

The first thing that jumps to mind for me as an alternate solution would be an (optional) git commit hash annotation at the top of your data migration file, like:
```typescript
// schema:b1ddc77eaafe73bd6b92d73709275dce40c5b0ee

async function nameUsers(prisma: Prisma.TransactionClient) {
  await prisma.user.updateMany({
    data: {
      name: "Name for users :)",
    },
  });
}
````

Your system would check out the Prisma schema for that commit (`b1ddc77eaafe73bd6b92d73709275dce40c5b0ee`), generate the Prisma client from that, and run the migration with it.

Of course the downside to this approach is that you'd need to include the `.git` repository in e.g. your Docker image, but that's not a dealbreaker for most people I'd say.

Just one idea. I think your current solution is fine as well, though if you want the 99.5% of people to use your package as well, I'd make it optional and show how to use it without copied Prisma schemas?

**Comment by Softjey** [comment]
>The first thing that jumps to mind for me as an alternate solution would be an (optional) git commit hash annotation at the top of your data migration file

That's an interesting trick with the commit hash annotation—I'll definitely think about it.

> I'd make it optional and show how to use it without copied Prisma schemas

Yes, I think your suggestion is correct that the schema should be optional. I'll proceed with implementing it that way. Thank you for the input!

**Comment by JVMartin** [comment]
Thanks for sharing your solution @Softjey, it's a clever one! I'm sorry for being dismissive of it before, I should have asked questions instead to understand your use case better

**Comment by Stadly** [comment]
Another idea I've thought about is to on app startup:

1. Compare the folder `prisma/migrations` and the table `_prisma_migrations` and create a list of the migrations that have not yet been applied.
2. Remove the migrations that have not yet been applied from the folder `prisma/migrations`.
3. For each migration that has not yet been applied:
    1. Move the migration back into `prisma/migrations`.
    2. Run `prisma migrate deploy`. This should only apply a single migration.
    3. Check if there is a `post.ts` (or `.sh` or `.js`) in the migration folder. If so, run it.

I have not tested it, but I think it should be fairly easy. Any thoughts?

**Comment by Stadly** [comment]
> Another idea I've thought about is to on app startup:
> 
> 1. Compare the folder `prisma/migrations` and the table `_prisma_migrations` and create a list of the migrations that have not yet been applied.
> 2. Remove the migrations that have not yet been applied from the folder `prisma/migrations`.
> 3. For each migration that has not yet been applied:
>    
>    1. Move the migration back into `prisma/migrations`.
>    2. Run `prisma migrate deploy`. This should only apply a single migration.
>    3. Check if there is a `post.ts` (or `.sh` or `.js`) in the migration folder. If so, run it.
> 
> I have not tested it, but I think it should be fairly easy. Any thoughts?

Ah, now that I think about it, it of course doesn't solve the outdated Prisma Client problem 😖

**Comment by Softjey** [comment]
> 2. Remove the migrations that have not yet been applied from the folder `prisma/migrations`.
> 3. For each migration that has not yet been applied:
>    
>    1. Move the migration back into `prisma/migrations`.
>    2. Run `prisma migrate deploy`. This should only apply a single migration.
>    3. Check if there is a `post.ts` (or `.sh` or `.js`) in the migration folder. If so, run it.

My library works in this way, but with additional PrismaClient generation, so it's quite simple

**Comment by wmadden** [comment]
Hi @sarink,

Thank you for your suggestion. I'm closing this issue as a duplicate of https://github.com/prisma/prisma/issues/19513 since it has more upvotes. To help prioritize this feature, please upvote #19513 and follow the discussion there. If you believe your use case has unique aspects, feel free to comment on that issue. Thanks for your input!
